keps/sig-api-machinery/1693-warnings/README.md

```diff
@@ -109,38 +109,39 @@ In client-go:
    3. log warnings with klog.Warn
 3. Add a process-wide warning handler (defaulting to the klog.Warn handler)
 4. Add a per-client warning handler (defaulting to the process-wide warning handler)
-
+ 
 In kubectl, configure the per-process handler to:
 
 1. dedupe warnings (only print a given warning once per invocation)
 2. log to stderr with a `Warning:` prefix
 3. color the `Warning:` prefix if stderr is a terminal and `$TERM != "dumb"` and `$NO_COLOR` is unset
-
+ 
 In kube-apiserver and kube-controller-manager, configure the process-wide handler to ignore warnings
 
 ## Design Details
-
+ 
 Server-side:
 
 * Add a handler chain filter that attaches a WarningRecorder implementation to the request context
-* Add a WarningRecord implementation that deduplicates the warning message per request and writes a `Warning` header
+* Add a WarningRecorder implementation that de-duplicates the warning message per request and writes a `Warning` header
   * the header structure is defined in [RFC2616#14.46](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.46)
   * warnings would be written with code `299` and warn-agent of `-`
 * In the endpoint installer, decorate handlers for deprecated APIs:
   * add the warning header
-  * increment the counter for the deprecated use metric
-  * add an audit annotation indicating the request was to a deprecated API
-
+  * increment the counter metric for the deprecated use
+  * add an audit annotation indicating the request was made to a deprecated API
+ 
 Client-side:
 
 * Parse `Warning` headers in server responses
-* Ignore malformed warning headers and warnings with codes other than `299`,
-  ensuring that this enhancement will not cause any previously successful API request to fail
+* Ignore malformed warning headers, ensuring that this enhancement 
+  will not cause any previously successful API request to fail
+* Ignore warnings with codes other than `299`
```

Comment:
Seems a little weird since some of the other warnings in the RFC can presented to a user even if we don't originate them.

Reply:
I could be talked into including `199` codes as well... 

> 1xx codes ... can only be generated by a cache when validating a cached entry, and MUST NOT be generated in any other situation.
>
> * Warning: 199 - "Miscellaneous Warning": The warning text can include arbitrary information to be presented to a human user or logged.  A system receiving this warning MUST NOT take any automated action, besides presenting the warning to the user.

I'm not as familiar or confident in use of cache-generated warnings. The 2xx warnings seemed clearer:

> * Warning: 214 - "Transformation Applied": This Warning code MUST be added by a proxy if it applies any transformation to the representation, such as changing the content-coding, media-type, or modifying the representation data, unless this Warning code already appears in the response.

This specifically means the content was transformed by a proxy. If that occurred, and the request still succeeded, it's not clear that presenting the warning to the user is useful (proxy transformation warnings certainly aren't the primary focus of this enhancement, or the likely field-level warning successors). I'm also not familiar enough with proxies to know how common this is, and wanted to be conservative in what we started exposing.

> * Warning: 299 - "Miscellaneous Persistent Warning": The warning text can include arbitrary information to be presented to a human user or logged.  A system receiving this warning MUST NOT take any automated action.

This seemed like the clearest code to surface - warnings not related to cache layers or proxy transformations, with explicit guidance to present to human users.

Reply:
updated to include 199 and 299 warnings